### PR TITLE
Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time [Merger of 89672 and 89773]

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90089
* __->__ #90039

After all of the preparatory commits, this is a subset of the
changes in https://github.com/pytorch/pytorch/pull/89392 that actually
change us to propagating fake tensors to backends.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

This is the merger of Ed's PR #89672, which is a rewrite of an older PR of mine (#89392), with CI Fixes on top of it (#89773)

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/dynamo/test_dynamic_shapes.py [hunk 1]
```diff
@@ -66,6 +66,7 @@ def make_dynamic_cls(cls):
     # Cannot call sizes() on tensor with symbolic sizes/strides
 )
 
+
```
Comment (ezyang) [inline test/dynamo/test_dynamic_shapes.py:69]: WHITESPAAAACE

File: test/dynamo/test_modules.py [hunk 2]
```diff
@@ -703,6 +704,9 @@ def test_generation_tag(self):
         m3 = deepcopy(m1)
         self.assertEqual(GenerationTracker.get_generation_value(m3), cur_generation)
 
+    # torch._subclasses.fake_tensor.UnsupportedFakeTensorException: meta converter nyi
+    # due to custom subclass (TensorProxy)
+    @unittest.expectedFailure
```
Comment (ezyang) [inline test/dynamo/test_modules.py:709]: I have a fix for this in a local patchset, this xfail AOK

File: test/dynamo/test_verify_correctness.py [hunk 3]
```diff
@@ -100,8 +100,11 @@ def compiler_fn(graph, example_inputs):
         r3 = opt_fn(a, (b, c), d)
 
         self.assertIsNotNone(r1)
-        self.assertTrue(same(r1, r2))
-        self.assertTrue(same(r1, r3))
```
Comment (ezyang) [inline test/dynamo/test_verify_correctness.py:104]: Why don't you just assert `same(r2, r3)` and call it a day lol

File: torch/fx/passes/shape_prop.py [hunk 4]
```diff
@@ -109,11 +109,36 @@ def forward(self, x):
...
+        self.real_module = self.module
+
     def run_node(self, n : Node) -> Any:
         try:
+            if self.fake_module is not None:
+                # Hacky little swap lol
+                self.module = self.fake_module
+
             result = super().run_node(n)
+            self.module = self.real_module
```
Comment (ezyang) [inline torch/fx/passes/shape_prop.py:141]: If run_node raises an exception this won't reset the module.

File: test/dynamo/test_verify_correctness.py (hunk 3 above)
Comment (voznesenskym) [reply test/dynamo/test_verify_correctness.py:104]: we care about r1 as well

File: torch/fx/passes/shape_prop.py (hunk 4 above)
Comment (ezyang) [reply torch/fx/passes/shape_prop.py:141]: There's no need to do a swap like this. The only thing you need to override is a few ops (`get_attr`, `call_module` I think?) to fetch from fake module rather than real module.

Comment (voznesenskym) [reply torch/fx/passes/shape_prop.py:141]: good call

File: torch/_dynamo/utils.py [hunk 5]
```diff
@@ -398,6 +398,10 @@ def clone_tensor(x):
 
 def clone_input(x):
     """copy while preserving strides"""
+    # TODO: this is questionable
+    if isinstance(x, torch._subclasses.FakeTensor):
+        # this func fails on fake tensors in __torch_dispatch__
+        return x
```
Comment (ezyang) [inline torch/_dynamo/utils.py:404]: I'll put up with this because I think my pending refactor will resolve the confusion here

File: torch/_dynamo/utils.py [hunk 6]
```diff
@@ -1155,3 +1165,20 @@ def assert_no_fake_params_or_buffers(gm):
...
+    Returns None if no fake_mode is fine
+    """
+    flat_inputs, _ = tree_flatten(inputs)
+    fake_mode = None
+    for flat_input in flat_inputs:
+        if isinstance(flat_input, torch._subclasses.FakeTensor):
+            if fake_mode is None:
+                fake_mode = flat_input.fake_mode
+            else:
+                assert fake_mode == flat_input.fake_mode
```
Comment (ezyang) [inline torch/_dynamo/utils.py:1183]: is, not equality, plz

File: torch/fx/passes/shape_prop.py (hunk 4 above)
Comment (voznesenskym) [reply torch/fx/passes/shape_prop.py:141]: > If run_node raises an exception this won't reset the module.

good call

> There's no need to do a swap like this. The only thing you need to override is a few ops (`get_attr`, `call_module` I think?) to fetch from fake module rather than real module.

Yeah but we already override this, this feels simpler. No strong opinion tho.

File: torch/_dynamo/optimizations/backends.py [hunk 7]
```diff
@@ -747,9 +747,16 @@ def torchxla_trivial(subgraph):
...
     import torch._dynamo.optimizations.torchxla_integration as integration
 
-    model = subgraph.model
-    example_inputs = subgraph.example_inputs
-    return integration.extract_compiled_graph(model, example_inputs)
+    compiled_graph = None
+
+    def fwd(*args):
+        nonlocal compiled_graph
+        model = subgraph.model
```
Comment (ezyang) [inline torch/_dynamo/optimizations/backends.py:754]: nit: this will keep subgraph permanently alive, whereas previously it could have been GC'ed after compilation. You should del subgraph being done with compilation

File: torch/_dynamo/optimizations/distributed.py [hunk 8]
```diff
@@ -211,7 +214,7 @@ def __init__(self, module, compiler):
                 super().__init__(module)
                 self.compiler = compiler
 
-            def compile_submod(self, submod, args, kwargs):
+            def compile_submod(self, input_mod, args, kwargs):
```
Comment (ezyang) [inline torch/_dynamo/optimizations/distributed.py:217]: The variable renaming makes it harder for code reviewers

Comment (ezyang) [reply torch/_dynamo/optimizations/distributed.py:217]: like, afaict, there's literally no change from lines 211 to 250 but I had to carefully audit to make sure there weren't typos lol

Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:217]: This is from your PR?

File: torch/_dynamo/optimizations/distributed.py [hunk 9]
```diff
@@ -236,42 +239,61 @@ def forward(self, *args):
...
                 wrapper = WrapperModule(
-                    self.compiler(submod, args),
+                    self.compiler(input_mod, args),
                     unwrap_singleton_tuple,
                 )
                 return wrapper
 
             def run_node(self, n: Node) -> Any:
                 with fx_traceback.append_stack_trace(n.stack_trace):
                     args, kwargs = self.fetch_args_kwargs_from_env(n)
```
Comment (ezyang) [inline torch/_dynamo/optimizations/distributed.py:257]: It would be really nice to have a comment here explaining what's going on

File: torch/_dynamo/optimizations/distributed.py [hunk 10]
```diff
@@ -236,42 +239,61 @@ def forward(self, *args):
...
+                    new_args = []
+                    assert fake_mode
+                    for arg in args:
+                        if isinstance(arg, torch.Tensor) and not isinstance(
+                            arg, torch._subclasses.FakeTensor
+                        ):
+                            new_args.append(fake_mode.from_tensor(arg))
```
Comment (ezyang) [inline torch/_dynamo/optimizations/distributed.py:264]: It would be much better if we could assume all the args are already fakeified (and we just maintain the invariant that all the intermediates are fake tensors). In particular, fakeifying a tensor can trigger the allocation of new symints, but if they're not properly associated as inputs we may not be able to instantiate those variables on subsequent runs.

tl;dr I suspect arg is always a fake tensor here

File: torch/_dynamo/optimizations/distributed.py [hunk 11]
```diff
@@ -236,42 +239,61 @@ def forward(self, *args):
...
                     if n.op == "call_module":
-                        submod = self.fetch_attr(n.target)
-                        log.debug(f"\n---{n.target} graph---\n" + str(submod.graph))
-                        compiled_submod = self.compile_submod(submod, args, kwargs)
+                        real_mod = self.fetch_attr(n.target)
+                        if fake_mode:
+                            fake_submod = deepcopy_to_fake_tensor(real_mod, fake_mode)
+                        else:
+                            fake_submod = real_mod
+                            pass
```
Comment (ezyang) [inline torch/_dynamo/optimizations/distributed.py:280]: pass here unnecessary

File: test/dynamo/test_modules.py (hunk 2 above)
Comment (voznesenskym) [reply test/dynamo/test_modules.py:709]: k

File: test/dynamo/test_dynamic_shapes.py (hunk 1 above)
Comment (voznesenskym) [reply test/dynamo/test_dynamic_shapes.py:69]: k

File: torch/_dynamo/output_graph.py [hunk 12]
```diff
@@ -88,21 +88,10 @@ def __repr__(self):
...
-
-        wrapper_backend_compiler_fn._torchdynamo_orig_callable = compiler_fn  # type: ignore[attr-defined]
-        return wrapper_backend_compiler_fn
-
-    return compiler_fn
-
-
 class WrapperBackend:
-    def __init__(self, backend: CompilerFn):
+    def __init__(self, backend, original_example_inputs):
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:92]: y u delete the type annotation lol

File: torch/_dynamo/optimizations/backends.py (hunk 7 above)
Comment (voznesenskym) [reply torch/_dynamo/optimizations/backends.py:754]: good idea

File: torch/_dynamo/optimizations/distributed.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:257]: Sure thing

File: torch/_dynamo/output_graph.py [hunk 13]
```diff
@@ -541,16 +529,42 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
...
 
+    def fake_example_inputs(self) -> List[torch.Tensor]:
+        result = []
+        for arg in self.graphargs:
+            example = arg.get_fake_examples()
+            if example is not None:
+                result.extend(example)
+            else:
+                # Fallback, in case fake_tensor was not set
+                # Particularly for graph args that are not tensors
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:564]: I hope graph args can't be None, because otherwise there is type confusion here.

File: torch/_dynamo/optimizations/distributed.py (hunk 10 above)
Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:264]: Im not sure we can, I'm pretty sure we saw non fakes in here, but maybe that was in an intermediate state of this PR. I can check again.

File: torch/_dynamo/output_graph.py [hunk 14]
```diff
@@ -541,16 +529,42 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
...
+            # WrapperBackend needs real inputs, for now, to verify correctness
             if config.verify_correctness:
-                compiler_fn = wrap_compiler_fn(compiler_fn)
-            compiled_fn = compiler_fn(gm, self.example_inputs())
+                compiler_fn = WrapperBackend(compiler_fn, self.example_inputs())
+
+            # We invoke the backend with real inputs if and only if
+            # we are in doing accuracy evaluation in the minifier
+            # Note: This check is the same as in `save_graph_repro`
+            is_minifier_backend = "_accuracy" in str(
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:539]: nuh uh, we're not testing for minifier backend this way. Why can't you just do an `is` test here, there's what, one function you need to test?

File: torch/_dynamo/output_graph.py [hunk 15]
```diff
@@ -541,16 +529,42 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
...
+                torch._dynamo.eval_frame.innermost_fn(compiler_fn)
+            )
+            is_top_level_minifiying = (
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:542]: nit: minifying

File: torch/_dynamo/variables/builder.py [hunk 16]
```diff
@@ -108,8 +108,13 @@ class GraphArg:
     source: Source
     example: Any
     is_unspecialized: bool
+    fake_tensor: Optional[torch._subclasses.fake_tensor.FakeTensor]
 
     def __post_init__(self):
+        if isinstance(self.example, torch.Tensor):
+            assert self.fake_tensor is not None and isinstance(
+                self.fake_tensor, torch._subclasses.fake_tensor.FakeTensor
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:116]: you don't need to do the is None test, with the isinstance test

File: torch/_dynamo/variables/builder.py [hunk 17]
```diff
@@ -689,11 +713,14 @@ def _clone_input(value):
     with preserve_rng_state():
         if example_value is None:
             example_value = get_fake_value(proxy.node, tx)
-
         else:
-            proxy.tracer.real_value_cache[proxy.node] = _clone_input(example_value)
-            fake_wrapper = functools.partial(wrap_to_fake_tensor_and_record, tx=tx)
-            example_value = fake_wrapper(example_value)
+            # Note: Unfortunately, this can happen during tracing, and is valid enough for now to allow.
+            # TODO(voz): Find all the callsites and burn this down.
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:718]: I think I also managed to nail a bunch of these in pending PR

File: torch/_dynamo/variables/builder.py [hunk 18]
```diff
@@ -551,10 +565,7 @@ def wrap_tensor(self, value: torch.Tensor):
                 # guards=self.make_guards(GuardBuilder.TENSOR_MATCH),
             )
         else:
-            if not is_constant_source(self.get_source()):
-                self.tx.output.graphargs.append(
-                    GraphArg(self.get_source(), value, False)
-                )
+            graph_arg = None
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:568]: this None assignment is unnecessary

File: test/dynamo/test_optimizations.py [hunk 19]
```diff
@@ -120,6 +120,38 @@ def compiler_fn(graph, example_inputs):
...
+                nonlocal r1
+                r = graph.forward(*args)
+                r1 = r[0]
+                return r
+
+            return fwd
+
+        a = torch.empty(2).fill_(1)
+        b = torch.empty(2).fill_(2)
+        c = torch.empty(2).fill_(3)
```
Comment (ezyang) [inline test/dynamo/test_optimizations.py:148]: The code duplication here is a bit irritating. Is there something in particular you are leaving `test_example_inputs` around to test?

File: torch/_dynamo/test_minifier_common.py [hunk 20]
```diff
@@ -100,13 +99,14 @@ def _run_repro(self, patch_code, repro_dir):
...
         )
-
         return repro_proc, repro_code
 
     # Template for testing code.
     # `run_code` is the code to run for the test case.
     # `patch_code` is the code to be patched in every generated file.
-    def _gen_test_code(self, run_code, repro_after, repro_level, patch_code):
+    def _gen_test_code(
+        self, run_code, repro_after, repro_level, patch_code, fake_tensor
```
Comment (ezyang) [inline torch/_dynamo/test_minifier_common.py:108]: Do you even use `fake_tensor`?

File: torch/_dynamo/optimizations/distributed.py (hunk 8 above)
Comment (ezyang) [reply torch/_dynamo/optimizations/distributed.py:217]: which is originally from https://github.com/pytorch/pytorch/pull/89392/files#diff-ebfc1054a8736abfc050c0a51c1f64f323a8cc3088f3f621d628cf7b421401faR219 lol

File: torch/_dynamo/optimizations/distributed.py (hunk 10 above)
Comment (ezyang) [reply torch/_dynamo/optimizations/distributed.py:264]: you can check this by pushing a separate PR that tightens the invariant and see if it fails or not

File: test/dynamo/test_optimizations.py (hunk 19 above)
Comment (voznesenskym) [reply test/dynamo/test_optimizations.py:148]: A backend that returns straight from example input should have fakes but be correct.

File: torch/_dynamo/debug_utils.py [hunk 21]
```diff
@@ -762,6 +762,8 @@ def backend_accuracy_fails(gm, example_inputs, compiler_fn, only_fwd=False):
 
 backend_aot_accuracy_fails = functools.partial(backend_accuracy_fails, only_fwd=True)
 
+MINIFIER_SPAWNED = False
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:765]: Please add comment here, explaining why this exists, how it is used, and when we would be able to remove it

File: torch/_dynamo/optimizations/distributed.py [hunk 22]
```diff
@@ -236,42 +239,75 @@ def forward(self, *args):
...
                 wrapper = WrapperModule(
-                    self.compiler(submod, args),
+                    self.compiler(input_mod, args),
                     unwrap_singleton_tuple,
                 )
                 return wrapper
 
+            # Note:
+            #
+            # The way distributed works today we fakify the module for compile time,
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_dynamo/optimizations/distributed.py:257]: The confusion here is what is meant by "compile time", which can be reasonably miscontrued to mean "what I pass to the compiler function" whereas what you actually mean is "compile time fake tensor propagation, so we can pass fake tensors into the submodule compiles"

File: torch/_dynamo/output_graph.py [hunk 23]
```diff
@@ -541,16 +529,38 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
...
             _step_logger()(logging.INFO, f"calling compiler function {name}")
             compiler_fn = self.compiler_fn
+            # WrapperBackend needs real inputs, for now, to verify correctness
             if config.verify_correctness:
-                compiler_fn = wrap_compiler_fn(compiler_fn)
-            compiled_fn = compiler_fn(gm, self.example_inputs())
+                compiler_fn = WrapperBackend(compiler_fn, self.example_inputs())
+
+            # We invoke the backend with real inputs if and only if
+            # we are in doing accuracy evaluation in the minifier
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:537]: Comment for why this is a hack and when it can be removed

**Review by ezyang — APPROVED** [review]

File: torch/_dynamo/optimizations/distributed.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:257]: I'll rewrite it.

**Comment by voznesenskym** [comment]
> @pytorchbot revert -m "broke xla tests https://hud.pytorch.org/pytorch/pytorch/commit/ef0c7ec958439caf44a98fb7b70d920c6c2264b9 https://github.com/pytorch/pytorch/actions/runs/3606308473/jobs/6077646142" -c landrace

I'd rather we have forward fixed, its a 1 liner to fix. Alas.

**Comment by ezyang** [comment]
the way to correctly land xla changes is to update the hash in the PR to your branch on xla. Once landed, merge the xla branch into master. The automatic head updating process will eventually reset xla's commit id to master.

**Review by ezyang — APPROVED** [review]

**Comment by voznesenskym** [comment]
> the way to correctly land xla changes is to update the hash in the PR to your branch on xla. Once landed, merge the xla branch into master. The automatic head updating process will eventually reset xla's commit id to master.

Yes. But I did it in a way that should not need XLA changes.
